src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleDictionary.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonSerializer
+    {
+        private static void HandleStartDictionary(JsonSerializerOptions options, ref Utf8JsonReader reader, ref ReadStack state)
+        {
+            Debug.Assert(!state.Current.IsProcessingEnumerable);
+
+            JsonPropertyInfo jsonPropertyInfo = state.Current.JsonPropertyInfo;
+            if (jsonPropertyInfo == null)
+            {
+                jsonPropertyInfo = state.Current.JsonClassInfo.CreateRootObject(options);
+            }
+
+            Debug.Assert(jsonPropertyInfo != null);
+
+            // A nested object or dictionary so push new frame.
+            if (state.Current.PropertyInitialized)
+            {
+                Debug.Assert(state.Current.IsDictionary);
+
+                JsonClassInfo classInfoTemp = state.Current.JsonClassInfo;
+                state.Push();
+                state.Current.JsonClassInfo = classInfoTemp.ElementClassInfo;
+                state.Current.InitializeJsonPropertyInfo();
+
+                ClassType classType = state.Current.JsonClassInfo.ClassType;
+                if (classType == ClassType.Value &&
+                    jsonPropertyInfo.ElementClassInfo.Type != typeof(object) &&
+                    jsonPropertyInfo.ElementClassInfo.Type != typeof(JsonElement))
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(state.Current.JsonClassInfo.Type, reader, state.PropertyPath);
```

Comment:
When will we hit this? I thought we already validated this earlier.

Reply:
The validation previously doesn't cover cases from the root like this: `JsonSerializer.Parse<Dictionary<string, string>>`.

It is possible to extend the "PropertyPolicy" mechanism to also catch this, however. Outside the scope of this PR.